go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,551 @@
+		_, err = hacluster.cluster.Kube.CreateEndpoint(config.Kubernetes.OVNConfigNamespace, &ovndbEp)
+		if err != nil {
+			logrus.Errorf("%s Endpount Create failed", ovnkubeDbEp)
+		}
+	} else {
+		logrus.Debugf("updateOvnDbEndpoints : Updating the endpoint")
+		ovndbEp := ep.DeepCopy()
+		ovndbEp.Subsets = epSubsets
+
+		epAnnotations := ovndbEp.ObjectMeta.GetAnnotations()
+		epAnnotations["ovnkube-master-leader"] = hacluster.nodeName
+		ovndbEp.ObjectMeta.SetAnnotations(epAnnotations)
+		_, err := hacluster.cluster.Kube.UpdateEndpoint(config.Kubernetes.OVNConfigNamespace, ovndbEp)
+		if err != nil {
+			logrus.Errorf("%s Endpount Update failed", ovnkubeDbEp)
+		}
+	}
+	return err
+}
+
+// ConfigureAsStandby configures the node as standby
+func (hacluster *OvnHAMasterController) ConfigureAsStandby(ep *kapi.Endpoints) error {
+	if !hacluster.manageDBServers {
+		// Nothing to do if not managing db servers.
+		return nil
+	}
+
+	// Get the master ip
+	masterIPList, sbDBPort, nbDBPort, err := extractDbRemotesFromEndpoint(ep)
+	if err != nil {
+		// The db remotes are invalid. Return without doing anything.
+		// Once master updates the endpoints properly we will be notified.
+		logrus.Infof("ConfigureAsStandby : error in extracting DbRemotes From Endpoint")
+		return nil
+	}
+
+	logrus.Infof("ConfigureAsStandby: New leader IP is : [%s]", masterIPList[0])
+
+	// Get the active ovsdb-server address.
+	var stdout string
+	stdout, _, err = util.RunOVNNBAppCtl("ovsdb-server/get-active-ovsdb-server")
+	if err != nil {
+		logrus.Errorf("Getting  active-ovsdb-server of NB ovsdb-server failed")
+		return err
+	}
+
+	activeServerOutOfSync := false
+	s := strings.Split(stdout, ":")
+	if len(s) != 3 {
+		activeServerOutOfSync = true
+	} else {
+		if s[0] != "tcp" || s[1] != masterIPList[0] || s[2] != strconv.Itoa(int(nbDBPort)) {
+			activeServerOutOfSync = true
+		}
+	}
+
+	if !activeServerOutOfSync {
+		stdout, _, err = util.RunOVNSBAppCtl("ovsdb-server/get-active-ovsdb-server")
+		if err != nil {
+			logrus.Errorf("Getting  active-ovsdb-server of SB ovsdb-server failed")
+			return err
+		}
+
+		s := strings.Split(stdout, ":")
+		if len(s) != 3 {
+			activeServerOutOfSync = true
+		} else {
+			if s[0] != "tcp" || s[1] != masterIPList[0] || s[2] != strconv.Itoa(int(sbDBPort)) {
+				activeServerOutOfSync = true
+			}
+		}
+	}
+
+	if activeServerOutOfSync {
+		logrus.Debugf("ConfigureAsStandby : active server out of sync..Setting the new active server to : " + masterIPList[0])
+		err = hacluster.DemoteOVNDbs(masterIPList[0], nbDBPort, sbDBPort)
+		if err != nil {
+			logrus.Errorf("Demoting OVN ovsdb-servers to standby failed")
+			return err
+		}
+	}
+	return nil
+}
+
+func (hacluster *OvnHAMasterController) validateOvnDbEndpoints(ep *kapi.Endpoints) bool {
+	if ep.Name != ovnkubeDbEp {
+		return false
+	}
+
+	epAnnotations := ep.ObjectMeta.GetAnnotations()
+	leader, present := epAnnotations["ovnkube-master-leader"]
```

Comment:
@numansiddique same here, should just be able to use `leader, present := ep.Annotations["ovnkube-master-leader"]

Reply:
Actually though, do we need the annotation? Won't it be sufficient to ensure that the port #s and the IP address are what this new master thinks they should be?

Reply:
Actually it is not really required. It's more of a book keeping. An admin can figure out who is the present leader by looking into the endpoint object.
Will this be useful ? Or shall i remove this code to save the leader info in the annotation ?

Reply:
Since any way  the code in L263 when creating or updating the endpoint stores this information, the code here just validates to be sure that it is as expected. If not it updates the endpoint.
Let me know if you want to remove and I will remove it.

Reply:
@numansiddique I think it's fine for now